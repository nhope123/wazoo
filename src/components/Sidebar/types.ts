export interface Stream {
  _id: string;
  viewers: number;
  channel: {
    url: string;
    status: string;
    display_name: string;
    views: number;
    followers: number;
    logo: string;
    banner: string | null;
  };
  game: string;
}

export interface Links {
  channel: string;
}

export interface ChannelData {
  _links: Links;
  stream: Stream | null;
}

export interface ChannelDetail {
  id: string;
  online: boolean;
  name: string;
  link: string;
  live_viewers?: number;
  status?: string;
  game?: string;
  views?: number;
  followers?: number;
  logo?: string;
  banner?: string;
}
