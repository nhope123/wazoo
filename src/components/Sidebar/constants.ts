const userList = [
  'ESL_SC2',
  'OgamingSC2',
  'cretetion',
  'freecodecamp',
  'storbeck',
  'habathcx',
  'RobotCaleb',
  'noobs2ninjas',
];
const TWITCH_URL =
  'https://twitch-proxy.freecodecamp.rocks/twitch-api/streams/';

export { userList, TWITCH_URL };