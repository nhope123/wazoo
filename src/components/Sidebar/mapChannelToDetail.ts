import { v4 as uid } from 'uuid';
import { ChannelData, ChannelDetail } from './types';

const defaultBanner =
  'https://s3.envato.com/files/156884535/Game_Background_Emerald_Lake_4270x2135.jpg';

const mapChannelToDetail = (user_data: ChannelData): ChannelDetail => {
  if (!user_data.stream) {
    const channel = user_data._links.channel.split('/');
    const name = channel[channel.length - 1];
    return { online: false, name, link: user_data._links.channel, id: uid() };
  }

  const { stream } = user_data;
  const { channel } = stream;
  return {
    id: stream._id,
    live_viewers: stream.viewers,
    link: channel.url,
    online: true,
    status: channel.status,
    name: channel.display_name,
    game: stream.game,
    views: channel.views,
    followers: channel.followers,
    logo: channel.logo,
    banner: channel.banner || defaultBanner,
  };
};

export default mapChannelToDetail;
