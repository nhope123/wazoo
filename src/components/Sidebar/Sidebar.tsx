import axios from 'axios';
import { FC, useCallback, useEffect, useMemo, useState } from 'react';
import { ChannelDetail, ChannelFilter, SidebarProps } from './types';
import Drawer from '@mui/material/Drawer';
import SidebarContent from './components/SearchFilter/SidebarContent';
import { useTheme } from '@mui/material/styles';
import useMediaQuery from '@mui/material/useMediaQuery';
import mapChannelToDetail from './helpers/mapChannelToDetail';

const userList = [
  'ESL_SC2',
  'OgamingSC2',
  'cretetion',
  'freecodecamp',
  'storbeck',
  'habathcx',
  'RobotCaleb',
  'noobs2ninjas',
];
const TWITCH_URL =
  'https://twitch-proxy.freecodecamp.rocks/twitch-api/streams/';

const Sidebar: FC<SidebarProps> = (props) => {
  const { isSidebarOpen, setChannel, setSidebarOpen, selectedChannel } = props;
  const [channels, setChannels] = useState<ChannelDetail[]>([]);
  const [filter, setFilter] = useState<ChannelFilter>('All');
  const [search, setSearch] = useState<string>('');

  const theme = useTheme();
  const isMobile = useMediaQuery(theme.breakpoints.down('md'));

  useEffect(() => {
    if (!isMobile && isSidebarOpen) {
      setSidebarOpen(false);
    }
  }, [isMobile, isSidebarOpen, setSidebarOpen, theme.breakpoints]);

  useEffect(() => {
    const data: ChannelDetail[] = [];
    const raw: any[] = [];

    Promise.all(userList.map((username) => axios(`${TWITCH_URL}${username}`)))
      .then((res) => {
        res.forEach((response) => {
          raw.push(response.data);
          data.push(mapChannelToDetail(response.data));
        });
        console.log('data', data);

        setChannels(data);
        setChannel(data[0]);
      })
      .catch((err) => console.log(`error: ${err}`));
  }, [setChannel]);

  const visibleChannels = useMemo(() => {
    let result: ChannelDetail[] = channels;
    if (filter !== 'All') {
      result = channels.filter((channel) =>
        filter === 'Online' ? channel.online : !channel.online,
      );
    }

    if (!search) return result;
    return result.filter((channel) =>
      channel.name.toLowerCase().includes(search.toLowerCase()),
    );
  }, [channels, filter, search]);

  const handleSearch = useCallback((search: string = '') => {
    setSearch(search);
  }, []);

  const handleChannelSelected = useCallback(
    (selection: ChannelDetail) => {
      setChannel(selection);
      if (isSidebarOpen) setSidebarOpen(false);
    },
    [isSidebarOpen, setChannel, setSidebarOpen],
  );

  return (
    <>
      <SidebarContent
        filter={filter}
        setFilter={setFilter}
        search={search}
        setSearch={handleSearch}
        selectedChannel={selectedChannel}
        channels={visibleChannels}
        setSelectedChannel={setChannel}
        sx={{ display: { xs: 'none', md: 'block' } }}
      />

      <Drawer
        anchor="left"
        open={isSidebarOpen}
        onClose={() => setSidebarOpen(false)}
        PaperProps={{
          sx: {
            mt: 7,
            borderRadius: '0 4px 0 0',
            maxWidth: (theme) => theme.spacing(43),
            width: '80%',
            backgroundColor: 'background.paper',
          },
        }}
        sx={{ display: { xs: 'block', md: 'none' } }}
      >
        <SidebarContent
          filter={filter}
          setFilter={setFilter}
          search={search}
          setSearch={handleSearch}
          selectedChannel={selectedChannel}
          channels={visibleChannels}
          setSelectedChannel={handleChannelSelected}
        />
      </Drawer>
    </>
  );
};

export default Sidebar;
