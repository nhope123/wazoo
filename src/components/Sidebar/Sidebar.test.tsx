import Sidebar from './Sidebar';
import { beforeEach, describe, expect, it, vi, Mocked } from 'vitest';
import { render, screen, fireEvent, waitFor } from '../../test/vitest-setup';
import { cleanChannelOnline, rawChannels } from '../../test/testData';
import axios from 'axios';

vi.mock('axios');

const mockSetChannel = vi.fn();
const mockSetSidebarOpen = vi.fn();

const renderSidebar = (props = {}) => {
  const defaultProps = {
    isSidebarOpen: true,
    setChannel: mockSetChannel,
    setSidebarOpen: mockSetSidebarOpen,
    selectedChannel: '',
    ...props,
  };

  return render(<Sidebar {...defaultProps} />);
};

vi.mock('./helpers/mapChannelToDetail', () => ({
  __esModule: true,
  default: () => cleanChannelOnline,
}));

const mockedAxios = axios as Mocked<typeof axios>;

describe('Sidebar', () => {
  beforeEach(() => {
    vi.clearAllMocks();
    
    // axios.get = vi.fn().mockResolvedValue(rawChannels);
  });

  it('should match snapshot', () => {
    const { asFragment } = renderSidebar();
    expect(asFragment()).toMatchSnapshot();
  });

  it('should close sidebar on mobile view', () => {
    window.innerWidth = 500;
    window.dispatchEvent(new Event('resize'));
    renderSidebar({ isSidebarOpen: true });

    expect(mockSetSidebarOpen).toHaveBeenCalledWith(false);
  });

  it('should fetch data and update state', async () => {
    mockedAxios.get.mockImplementation(() => Promise.resolve({ data: rawChannels }));
    renderSidebar();

    await waitFor(() => {
      expect(mockSetChannel).toHaveBeenCalled();
    });
  });

  it('should filter channels based on filter and search state', () => {
    const channels = [
      { name: 'ESL_SC2', online: true },
      { name: 'OgamingSC2', online: false },
    ];

    renderSidebar({ channels });

    fireEvent.change(screen.getByPlaceholderText('Search'), {
      target: { value: 'ESL' },
    });

    expect(screen.getByText('ESL_SC2')).toBeInTheDocument();
    expect(screen.queryByText('OgamingSC2')).not.toBeInTheDocument();
  });

  it('should update search state on handleSearch', () => {
    renderSidebar();

    fireEvent.change(screen.getByPlaceholderText('Search'), {
      target: { value: 'test' },
    });

    expect(screen.getByDisplayValue('test')).toBeInTheDocument();
  });

  it('should update selected channel and close sidebar on handleChannelSelected', () => {
    const channel = { name: 'ESL_SC2', online: true };

    renderSidebar({ isSidebarOpen: true });

    fireEvent.click(screen.getByText('ESL_SC2'));

    expect(mockSetChannel).toHaveBeenCalledWith(channel);
    expect(mockSetSidebarOpen).toHaveBeenCalledWith(false);
  });

  it('should display all channels when filter is set to "All"', () => {
    const channels = [
      { name: 'ESL_SC2', online: true },
      { name: 'OgamingSC2', online: false },
    ];

    renderSidebar({ channels });

    fireEvent.change(screen.getByPlaceholderText('Search'), {
      target: { value: '' },
    });

    expect(screen.getByText('ESL_SC2')).toBeInTheDocument();
    expect(screen.getByText('OgamingSC2')).toBeInTheDocument();
  });

  it('should display only online channels when filter is set to "Online"', () => {
    const channels = [
      { name: 'ESL_SC2', online: true },
      { name: 'OgamingSC2', online: false },
    ];

    renderSidebar({ channels });

    fireEvent.click(screen.getByText('Online'));

    expect(screen.getByText('ESL_SC2')).toBeInTheDocument();
    expect(screen.queryByText('OgamingSC2')).not.toBeInTheDocument();
  });

  it('should display only offline channels when filter is set to "Offline"', () => {
    const channels = [
      { name: 'ESL_SC2', online: true },
      { name: 'OgamingSC2', online: false },
    ];

    renderSidebar({ channels });

    fireEvent.click(screen.getByText('Offline'));

    expect(screen.queryByText('ESL_SC2')).not.toBeInTheDocument();
    expect(screen.getByText('OgamingSC2')).toBeInTheDocument();
    });


  it('should filter channels based on filter and search state', () => {
    const channels = [
      { name: 'ESL_SC2', online: true },
      { name: 'OgamingSC2', online: false },
    ];

    renderSidebar({ channels });

    fireEvent.change(screen.getByPlaceholderText('Search'), {
      target: { value: 'ESL' },
    });

    expect(screen.getByText('ESL_SC2')).toBeInTheDocument();
    expect(screen.queryByText('OgamingSC2')).not.toBeInTheDocument();
  });

  it('should update search state on handleSearch', () => {
    renderSidebar();

    fireEvent.change(screen.getByPlaceholderText('Search'), {
      target: { value: 'test' },
    });

    expect(screen.getByDisplayValue('test')).toBeInTheDocument();
  });

  it('should update selected channel and close sidebar on handleChannelSelected', () => {
    const channel = { name: 'ESL_SC2', online: true };

    renderSidebar({ isSidebarOpen: true });

    fireEvent.click(screen.getByText('ESL_SC2'));

    expect(mockSetChannel).toHaveBeenCalledWith(channel);
    expect(mockSetSidebarOpen).toHaveBeenCalledWith(false);
  });
});
