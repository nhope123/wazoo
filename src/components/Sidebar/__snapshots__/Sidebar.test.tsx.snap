// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Sidebar > should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid2-root MuiGrid2-direction-xs-row MuiGrid2-grid-xs-4 css-1qjd5z-MuiGrid2-root"
    data-testid="sidebar"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-zo3717-MuiPaper-root-MuiCard-root"
      style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
    >
      <div
        class="MuiBox-root css-yd8sa2"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-cmpglg-MuiFormControl-root-MuiTextField-root"
        >
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedStart css-3ot9ek-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium css-1gtl5f-MuiSvgIcon-root"
              data-testid="SearchRoundedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27c1.2-1.4 1.82-3.31 1.48-5.34-.47-2.78-2.79-5-5.59-5.34-4.23-.52-7.79 3.04-7.27 7.27.34 2.8 2.56 5.12 5.34 5.59 2.03.34 3.94-.28 5.34-1.48l.27.28v.79l4.25 4.25c.41.41 1.08.41 1.49 0s.41-1.08 0-1.49zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14"
              />
            </svg>
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedStart css-1q55ijt-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r1:"
              placeholder="Search users"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-w4cd9x"
              >
                <span
                  class="notranslate"
                >
                  â€‹
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiButtonGroup-root MuiButtonGroup-outlined MuiButtonGroup-horizontal MuiButtonGroup-fullWidth MuiButtonGroup-colorPrimary css-j0iy8w-MuiButtonGroup-root"
          role="group"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary MuiButtonGroup-firstButton css-i3fu1t-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiBadge-root css-63ezxx-MuiBadge-root"
            >
              All
              <span
                class="MuiBadge-badge MuiBadge-standard MuiBadge-invisible MuiBadge-anchorOriginTopRight MuiBadge-anchorOriginTopRightRectangular MuiBadge-overlapRectangular css-z28gz6-MuiBadge-badge"
              />
            </span>
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary MuiButtonGroup-middleButton css-1sfm6cb-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Online
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary MuiButtonGroup-lastButton css-1sfm6cb-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Offline
          </button>
        </div>
      </div>
      <ul
        class="MuiList-root MuiList-padding css-3zbplw-MuiList-root"
        data-testid="channel-list"
      />
    </div>
  </div>
</DocumentFragment>
`;
